datapath: Add a flag to set VXLAN and Geneve ports

errordeveloper opened the pull request:
This change moves port defaulting logic from the kernel into the
agent and exposes a flag that lets user set a custom port.
    
Defaults ports remain unchanged.

Towards: #15956

errordeveloper committed:
datapath: Add a flag to set VXLAN and Geneve ports

This change makes it possible for user to set a custom port
for VXLAN or Geneve. In order to anable that, the defaulting
logic was introduced in the agent, so kernel defaults are no
longer relied upon.

Default ports remain unchanged.

Towards: #15956

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

jibi commented on bpf/init.sh:
```diff
@@ -432,6 +434,9 @@ if [ "${TUNNEL_MODE}" != "<nil>" ]; then
 		ip link add name $ENCAP_DEV address $(rnd_mac_addr) type $TUNNEL_MODE external || encap_fail
 	}
 	ip link set $ENCAP_DEV mtu $MTU || encap_fail
+	if [ "${TUNNEL_PORT}" != "<nil>" ]; then
```
I'm not sure if this is a problem in practice, but would it be better to set the `dstport` directly in the command above (`ip link add name $ENCAP_DEV address $(rnd_mac_addr) type $TUNNEL_MODE external`) rather than changing it after creating the device? We could have `$TUNNEL_PORT` always set (to either the default value or the user specified one) and save this if case/setting it to `<nil>`

errordeveloper replied:
> would it be better to set the dstport directly in the command above (`ip link add name ...`)...

So actually the ideas is that port can change after device is created, so we need to cater for that. It maybe the case that `dstport` needs to be specified in both invocation, if we are worried about any race conditions here...

errordeveloper replied:
> We could have `$TUNNEL_PORT` always set (to either the default value or the user specified one) and save this if case/setting it to `<nil>`

It is actually always going to be set, when `TUNNEL_MODE` is either `vxlan` or `geneve`, it's only set to `<nil>` when `TUNNEL_MODE` is neither of those two options... I this check is mostly for the sake of explicitness.

jibi replied:
> It is actually always going to be set

Sorry, what I meant with that was: always explicitly set to a specific port number (so that init.sh doesn't have to deal with the `<nil>` case)

errordeveloper replied:
> so that init.sh doesn't have to deal with the `<nil>` case

This check is somewhat redundant, I could delete it now and behaviour shouldn't change. But as far as this script itself is concerned, `<nil>` case is still possible for `TUNNEL_PORT`, and that is because there is no check for exact value of `TUNNEL_MODE` within this specific block.

errordeveloper commented on install/kubernetes/cilium/README.md:
```diff
@@ -352,5 +352,6 @@ contributors across the globe, there is almost always someone available to help.
 | tls | object | `{"enabled":true,"secretsBackend":"local"}` | Configure TLS configuration in the agent. |
 | tolerations | list | `[{"operator":"Exists"}]` | Node tolerations for agent scheduling to nodes with taints ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/ |
 | tunnel | string | `"vxlan"` | Configure the encapsulation configuration for communication between nodes. Possible values:   - disabled   - vxlan (default)   - geneve |
+| tunnelPort | int | `8472` (VXLAN), `6081` (Geneve) | Configure the tunnel port to use |
```
@bmcustodio helm-docs doesn't seem to like this, what do you think I should do here? I want to document this value, but I don't want to set it to zero by default... neither I want to have separate values for VXLAN and Geneve, that just doesn't seem to make sense.

bmcustodio replied:
I am not entirely sure what you mean, but looking at https://github.com/cilium/cilium/pull/16874/files#diff-83e9b8b2ccacddb2fbdb28fbcfa148473d05056c0b4c955fc971709b469d7331R355 I would say you want to add `tunnelPort` as a value to `values.yaml` and document it there:

```yaml
# -- Configure the tunnel port to use.
# By default, this is the default port for VXLAN (the default value for `.tunnel`).
# You may want to set this value to 6081 if using Geneve instead.
tunnelPort: 8472
```

If you do this you probably want to remove the `{{- if hasKey .Values "tunnelPort" }}` check.

errordeveloper replied:
My intension was to leave the defaulting to the agent, I am avoid duplicating it in the Helm template. And setting `tunnelPort` here for the default value of `tunnel` is also not quite right, user should be able to use `tunnel: geneve` without worrying about the port.

bmcustodio replied:
That makes sense, but then I would not add `tunnelPort` to the chart — if it's up to the agent to figure out, then it doesn't _need_ to be in the configmap right? If that's the case, I would just leave `hasKey` to accomodate for the case where the user does want to customize that, but keep it as a "hidden" value (i.e. not listed in `values.yaml` or docs). We do that already for some other values AFAIU.

errordeveloper replied:
Ok, it's probably fine to keep this hidden, right now the use-case for it is to do with OpenShift, and we can document it in OpenShift docs.

errordeveloper commented on pkg/datapath/loader/base.go:
```diff
@@ -363,6 +364,15 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
+	case option.TunnelGeneve:
+		args[initArgTunnelPort] = fmt.Sprintf("%d", option.TunnelPortDefaultGeneve)
+	default:
+		args[initArgTunnelPort] = "<nil>"
```
@jibi did you see this? I'm just following the general patter of how undefined arguments are handled by this function.

errordeveloper commented on bpf/init.sh:
```diff
@@ -432,6 +434,9 @@ if [ "${TUNNEL_MODE}" != "<nil>" ]; then
 	}
 	ip link set $ENCAP_DEV mtu $MTU || encap_fail
+	if [ "${TUNNEL_PORT}" != "<nil>" ]; then
+		ip link set $ENCAP_DEV dstport $TUNNEL_PORT || encap_fail
```
It looks this won't work, despite that being non-evident from the source code (https://github.com/cilium/cilium/issues/15956#issuecomment-878342725):
```
2021-07-15T08:46:49.775523869Z level=warning msg="+ ENCAP_DEV=cilium_vxlan" subsys=datapath-loader
2021-07-15T08:46:49.775526869Z level=warning msg="+ ip link show cilium_vxlan" subsys=datapath-loader
2021-07-15T08:46:49.775530169Z level=warning msg="4: cilium_vxlan: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000" subsys=datapath-loader
2021-07-15T08:46:49.775533770Z level=warning msg="    link/ether da:9e:0d:f9:97:0f brd ff:ff:ff:ff:ff:ff" subsys=datapath-loader
2021-07-15T08:46:49.775536870Z level=warning msg="+ ip link set cilium_vxlan mtu 1500" subsys=datapath-loader
2021-07-15T08:46:49.775540170Z level=warning msg="+ '[' 8472 '!=' '<nil>' ']'" subsys=datapath-loader
2021-07-15T08:46:49.775543370Z level=warning msg="+ ip link set cilium_vxlan dstport 8472" subsys=datapath-loader
2021-07-15T08:46:49.775547171Z level=warning msg="Error: either \"dev\" is duplicate, or \"dstport\" is a garbage." subsys=datapath-loader
```

errordeveloper replied:
Looks like I just need to add `type vxlan` here also.
```
       ip link set { DEVICE | group GROUP }
               [ { up | down } ]
               [ type ETYPE TYPE_ARGS ]
```

errordeveloper replied:
The parse accepts the option, but it errors out:
```
2021-07-15T09:20:49.108884244Z level=warning msg="+ ip link set cilium_vxlan type vxlan dstport 8472" subsys=datapath-loader
2021-07-15T09:20:49.108888544Z level=warning msg="Error: Cannot change port." subsys=datapath-loader
```

errordeveloper replied:
Will try to recreate the interface on port change, in any case port change would require tunnel reestablishment, so for the user who choses to change the port a disruption may seem reasonable. I don't know if we might need to add a flag, or if recreation of the interface might have other side-effects, let's discuss that here :)

errordeveloper replied:
@borkmann is there anything that needs to change in the datapath in the even of interface re-creation? Just wondering in case this might had been unforeseen previously.

errordeveloper replied:
I've submitted [a patch to fix iproute2](https://patchwork.kernel.org/project/netdevbpf/patch/20210716153557.10192-1-errordeveloper@gmail.com/).

pchaigno commented on pkg/datapath/loader/base.go:
```diff
@@ -363,6 +364,15 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
+	case option.TunnelGeneve:
+		args[initArgTunnelPort] = fmt.Sprintf("%d", option.TunnelPortDefaultGeneve)
+	default:
+		args[initArgTunnelPort] = "<nil>"
```
Hm, is it me or is the value from `--tunnel-port` never used?

We can also get rid of the `<nil>` case by defining `args[initArgTunnelPort]` to have the default `option.TunnelPortDefaultXXX` value if `--tunnel-port` isn't set.

errordeveloper replied:
> Hm, is it me or is the value from `--tunnel-port` never used?

Haha, thanks! Well spotted ;)

errordeveloper replied:
> We can also get rid of the `<nil>` case by defining `args[initArgTunnelPort]` to have the default `option.TunnelPortDefaultXXX` value if `--tunnel-port` isn't set.

The point was set to `<nil>` in case if somehow an unknown tunnel type is being used, I'm not seeing checks for unknown values anywhere, so it seems theoretically possible...

errordeveloper replied:
Spoke too soon, there is this error condition: https://github.com/cilium/cilium/blob/fb85fee7001e7fcc9fc3466c4ff018fe787fe26c/pkg/option/config.go#L2202-L2204

pchaigno commented on pkg/option/config.go:
```diff
@@ -1086,6 +1088,13 @@ func GetTunnelModes() string {
+	// TunnelPortDefaultVXLAN is the default VXLAN port
+	TunnelPortDefaultVXLAN int = 8472
+	// TunnelPortDefaultGeneve is the default Geneve port
+	TunnelPortDefaultGeneve int = 6081
```
Those should live in `defaults.go`.

errordeveloper commented on bpf/init.sh:
```diff
@@ -428,11 +439,17 @@ fi
 	ip link set $ENCAP_DEV mtu $MTU || encap_fail
 
+	if [ "${TUNNEL_PORT}" != "<nil>" ]; then
+		ip link show $ENCAP_DEV | grep "dstport $TUNNEL_PORT" || {
```
Looks like this doesn't actually work, `ip link show` doesn't seem to display what `dstport` is being used, or any other driver-specific flags... I am actually not sure how is one supposed to check this. I would assume there is ought to be way to do it using netlink directly?

cc @borkmann @jrfastab

errordeveloper replied:
`ip -details link show` 🚀

errordeveloper committed:
iproute2: use new image that fixes tunnel port attributes

xref https://github.com/cilium/image-tools/pull/136
xref https://github.com/cilium/iproute2/pull/14

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper committed:
update cilium-{runtime,builder}

This is to incorporate new iproute2 image

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

joestringer reviewed (approved):
Github workflow changes LGTM. I didn't look at the rest of the PR.

pchaigno commented on bpf/init.sh:
```diff
@@ -428,11 +439,17 @@ fi
 	ip link set $ENCAP_DEV mtu $MTU || encap_fail
 
+	if [ "${TUNNEL_PORT}" != "<nil>" ]; then
+		ip -details link show $ENCAP_DEV | grep "dstport $TUNNEL_PORT" || {
```
This change is causing us to always recreate the `cilium_vxlan` device in the development VM (and in Runtime tests). That is because, in the development VM, we are still using the old version of our iproute2 fork, which doesn't include the changes to properly display `dstport`: https://github.com/cilium/packer-ci-build/blob/72c2c6410237a795ac47bb1d08b7e4f309e4d2ce/provision/ubuntu/install.sh#L6.

We should probably fix this in packer-ci-build before merging this PR, to ensure we don't end up with flake in the Runtime test suite (might be okay, but recreating the device is also the sort of things that seems likely to have side effects). @brb already said on Slack he would fix it.

errordeveloper replied:
packer-ci-build doesn't yet incorporate iproute2 changes for libbpf, so perhaps it's best if @brb pick this up!

errordeveloper replied:
https://github.com/cilium/packer-ci-build/pull/281

pchaigno reviewed (commented):
LGTM, with one small change needed in packer-ci-build (see below).

Also, a nit: the first commit should IMO be the last commit since the other two commits updating images are needed for the first commit to work. Doesn't affect compilation though :-)

Using the development VM, I tested a migration from UDP/8472 to UDP/8473. As expected, connectivity between nodes is disrupted until both sides of any communication are using the same port. Once that is the case, connectivity is restored.
